### PR TITLE
Add validation to `RecordBatch` for non-nullable fields containing null values

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-rs/issues/1888

# Rationale for this change
 
 

Allowing null values in non-nullable fields can lead to incorrect results and missed optimizations in downstream crates such as DataFusion

# What changes are included in this PR?

Add validation check when creating a RecordBatch

# Are there any user-facing changes?

Yes. Users may start getting errors if their schemas are not currently defined correctly.

### CODE REVIEW

**Comment by andygrove** [comment]
@alamb I think this PR has exposed a bug in dictionary IPC encoding. The checked-in `dictionary_generated` files seem to have an incorrect schema (declaring a column as non-nullable when it contains null values). The roundtrip tests fail for the same reason.

File: arrow/src/record_batch.rs [hunk 1]
```diff
@@ -138,6 +138,20 @@ impl RecordBatch {
                 )
             })?;
 
+        for (c, f) in columns.iter().zip(&schema.fields) {
+            if !f.is_nullable() && c.null_count() > 0 {
+                if f.name().len() == 0 {
+                    // hacky workaround for known issue with dictionary IPC encoding
+                    // https://github.com/apache/arrow-rs/issues/1892
```
Comment (andygrove) [inline arrow/src/record_batch.rs:145]: See https://github.com/apache/arrow-rs/issues/1892

File: arrow/src/record_batch.rs [hunk 2]
```diff
@@ -138,6 +138,20 @@ impl RecordBatch {
                 )
             })?;
 
+        for (c, f) in columns.iter().zip(&schema.fields) {
+            if !f.is_nullable() && c.null_count() > 0 {
+                if f.name().len() == 0 {
```
Comment (tustvold) [inline arrow/src/record_batch.rs:143]: is_empty() is normally preferred

File: arrow/src/record_batch.rs (hunk 1 above)
Comment (tustvold) [reply arrow/src/record_batch.rs:145]: Personally I'd rather an `if !condition` over an `if else` with an empty if body

**Review by tustvold — APPROVED** [review]
Nice, minor nits, big fan of this

**Review by alamb — APPROVED** [review]
Thanks @andygrove

**Comment by andygrove** [comment]
Failed with a Rust / C# test this time

```
==========================================================
Testing file duration
==========================================================
Traceback (most recent call last):
################# FAILURES #################
FAILED TEST: datetime Rust producing,  C# consuming
1 failures
  File "/arrow/dev/archery/archery/integration/runner.py", line 246, in _run_ipc_test_case
    run_binaries(producer, consumer, test_case)
  File "/arrow/dev/archery/archery/integration/runner.py", line 286, in _produce_consume
    consumer.stream_to_file(producer_stream_path, consumer_file_path)
  File "/arrow/dev/archery/archery/integration/tester_csharp.py", line 63, in stream_to_file
    self.run_shell_command(cmd)
  File "/arrow/dev/archery/archery/integration/tester.py", line 49, in run_shell_command
    subprocess.check_call(cmd, shell=True)
  File "/opt/conda/envs/arrow/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '/arrow/csharp/artifacts/Apache.Arrow.IntegrationTest/Debug/net6.0/Apache.Arrow.IntegrationTest --mode stream-to-file -a /tmp/tmpyt_o8mf0/b64aa0aa_generated_datetime.consumer_stream_as_file < /tmp/tmpyt_o8mf0/b64aa0aa_generated_datetime.producer_file_as_stream' returned non-zero exit status 1.
```

**Comment by andygrove** [comment]
Other builds are failing with the same issue. I wonder if it is related to https://github.com/apache/arrow/pull/13279 ?

**Review by liukun4515 — APPROVED** [review]
LGTM

File: arrow/src/record_batch.rs [hunk 3]
```diff
@@ -138,6 +138,19 @@ impl RecordBatch {
                 )
             })?;
 
+        for (c, f) in columns.iter().zip(&schema.fields) {
+            if !f.is_nullable() && c.null_count() > 0 {
+                // hacky workaround for known issue with dictionary IPC encoding
+                // https://github.com/apache/arrow-rs/issues/1892
```
Comment (viirya) [inline arrow/src/record_batch.rs:144]: #1893 should fix this issue.

**Comment by tustvold** [comment]
I think a rebase should clear the test failures related to IPC dictionaries

File: arrow/src/record_batch.rs (hunk 3 above)
Comment (andygrove) [reply arrow/src/record_batch.rs]: Thanks @viirya I have upmerged and removed this workaround

**Comment by alamb** [comment]
🎉

**Comment by alamb** [comment]
Marking this as an API change because it caused many errors in the datafusion tests -- see https://github.com/apache/arrow-datafusion/pull/2778 -- and I think it should be highlighted prominently in the release notes

**Pull request merged**